include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,1386 @@
+			ix = choseRight ? right:(ix+1u);
+			weight *= choseRight ? (1.0-w):w;
+		}
+		else 
+#endif //OP_BLEND
+		{
+#ifdef OP_SET_GEOM_NORMAL
+			if (op==OP_SET_GEOM_NORMAL)
+			{
+				instr_execute_SET_GEOM_NORMAL();
+			} else 
+#endif //OP_SET_GEOM_NORMAL
+#ifdef OP_BUMPMAP
+			if (op==OP_BUMPMAP)
+			{
+				mat2x4 srcs = instr_fetchSrcRegs(instr);
+				instr_execute_BUMPMAP(instr, srcs);
+			} else
+#endif //OP_BUMPMAP
+			{}
+			ix = ix + 1u;
+		}
+
+		instr = irr_glsl_MC_fetchInstr(stream.offset+ix);
+		op = instr_getOpcode(instr);
+	}
+
+	bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
+	params_t params = instr_getParameters(instr, bsdf_data);
+	//speculatively
+	const float ax = params_getAlpha(params);
+	const float ax2 = ax*ax;
+	const float ay = params_getAlphaV(params);
+	const float ay2 = ay*ay;
+	const mat2x3 ior = bsdf_data_decodeIoR(bsdf_data,op);
+	const bool is_bsdf = !op_isBRDF(op);
+	const vec3 refl = params_getReflectance(params);
+
+#ifndef NO_TWOSIDED
+	bool ts_flag = false;
+	handleTwosided_interactionOnly(ts_flag, instr);
+#endif //NO_TWOSIDED
+
+	const bool is_plastic = op==OP_PLASTIC;
+
+#ifdef OP_DIFFUSE
+
+#define OP_DIFFUSE_ALIAS OP_DIFFUSE
+
+#else
+
+#define OP_DIFFUSE_ALIAS 0xdeadbeefu
+
+#endif
+
+	float pdf;
+	vec3 rem;
+	uint ndf = instr_getNDF(instr);
+	irr_glsl_LightSample s;
+	irr_glsl_AnisotropicMicrofacetCache uf;
+
+	// the if/else statements below cannot be done as switch/case because it checks both `op` and `ndf`
+#ifdef OP_PLASTIC
+	float plastic_spec_weight;
+	if (is_plastic) {
+		vec3 fresnel = irr_glsl_fresnel_dielectric(ior[0],currInteraction.isotropic.NdotV);
+		float ws = dot(fresnel, CIE_XYZ_Luma_Y_coeffs);
+		bool choseDiffuse = u.z>=ws;
+		op = choseDiffuse ? OP_DIFFUSE_ALIAS : op;
+		plastic_spec_weight = ws;
+	}
+#endif //OP_PLASTIC
+
+#if defined(OP_DIFFUSE) || defined(OP_PLASTIC)
+	if (op==OP_DIFFUSE_ALIAS) {
+		s = irr_glsl_oren_nayar_cos_generate(currInteraction, u.xy, ax2);
+		uf = getSmoothMicrofacetCache(currInteraction.isotropic.NdotV, s.NdotL);
+		rem = refl*irr_glsl_oren_nayar_cos_remainder_and_pdf(pdf, s, currInteraction.isotropic, ax2);
+	} else 
+#endif //OP_DIFFUSE
+
+#ifdef OP_DIFFTRANS
+	if (op == OP_DIFFTRANS) {
+		s = irr_glsl_lambertian_transmitter_cos_generate(currInteraction, u);
+		uf = getSmoothMicrofacetCache(currInteraction.isotropic.NdotV, s.NdotL);//TODO?
+		rem = refl*irr_glsl_lambertian_transmitter_cos_remainder_and_pdf(pdf, s);
+	} else
+#endif //OP_DIFFTRANS
+
+#ifdef OP_THINDIELECTRIC
+	if (op == OP_THINDIELECTRIC) {
+		const vec3 luminosityContributionHint = CIE_XYZ_Luma_Y_coeffs;
+		s = irr_glsl_thin_smooth_dielectric_cos_generate(currInteraction, u, ior[0]*ior[0], luminosityContributionHint);
+		uf = getSmoothMicrofacetCache(currInteraction.isotropic.NdotV, s.NdotL);
+		rem = irr_glsl_thin_smooth_dielectric_cos_remainder_and_pdf(pdf, s, currInteraction.isotropic, ior[0]*ior[0], luminosityContributionHint);
+		pdf = 1.0;
+	} else
+#endif
+
```

Comment:
there's a far easier way to deal with the generator, it will be less divergent, less code and easier to read & understand.

You should tear into the microfacet transmission generators to see how it all comes together.

It's best if you intialized `out_rem` to 1.0 for non plastics and then multiplied factors onto it.

### First of all handle your reflective fresnel

If your BSDF is not diffuse (diffuse or difftrans) then you will need to treat it somewhat like a blend node. The way you choose plastic vs. diffuse right now is 100% OK except for variable partition/rescale that is missing.

You need an extra flag besides `out_generatorInstr` to discern which sub-generator was used.

If you choose the diffuse generator, then you need to premultiply or initialize the `out_rem` to the fresnel diffuse correction factor **multiplied by `rcpChoiceProb`**, then proceed to the diffuse generator path.

However for the specular generator, then you need to premultiply or initialize the `out_rem` to be the `rcpChoiceProb`, you'd then need to proceed as for a CONDUCTOR, except that you **compute the fresnel like for a dielectric.**

## All Non-Diffuse BSDF generators generate a half vector H first

So you should split your code for generating an `s` into two branches, diffuse vs. non-diffuse.

### Diffuse path

Doesn't matter if its DIFF or DIFFTRANS, you need to generate a point on a hemisphere.

If the material is DIFFTRANS then you use the `u.z` variable to flip the `L` direction with 50% probability

PDF is trivial to compute.

The microfacet cache, you should compute using our built-in function that takes an interaction and a light sample (specialized for the reflective case).

### The Specular Path

Use the correct generator to generate `H` given your NDF, it doesn't matter if the material is CONDUCTOR or DIELECTRIC. If the material is smooth then simply set `H=N` (but this backend has no smooth materials, it just has NDFs with 0 roughness, so this point is moot).

Now for the `out_rem` compute the G2/G1 function (there might be some useful metadate that a generator overload might return).

To compute the PDF you need to multiply it by NDF*G1 initially, there's one extra step in case of a dielectric material.

Compute the Fresnel of `VdotH`, but dont use it yet.

If the material cannot be Refractive then set a `refraction` boolean to false.

Otherwise if the material could be Refractive (dielectric) then choose between reflection and refraction based on the previously calculated Fresnel's luma coefficients, divide the `out_rem` and multiply the PDF by the probability of your choice respectively.

Multiply `out_rem` by the correct Frensel depending on `refraction` boolean (either `F` or `1-F`)

Now there's a handy function in our GLSL library which can construct a microfacet cache given a view interaction, a microfacet `H` vector, a refraction boolean a few extra infos like the refractive index. Use it.

That's it, you're done!